### PR TITLE
Munchausen dqn

### PR DESCRIPTION
**Pull request by AleShi94**
## Description

Added Munchausen loss to DQN. Tested on CartPole

### CODE REVIEW

**Review by KohlerHECTOR — APPROVED** [review]
Hello ! LGTM ! Maybe it would be nice to have a small dqn_vs_mdqn file to do a quick comparison of mdqn and dqn on cartpole !

**Comment by AleShi94** [comment]
> Hello ! LGTM ! Maybe it would be nice to have a small dqn_vs_mdqn file to do a quick comparison of mdqn and dqn on cartpole !

Good idea, though I am thinking to do it later in a separate pull request?

File: rlberry/agents/torch/dqn/mdqn.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    model_config = size_model_config(env, **model_config)
+    return model_factory(**model_config)
+
+
+class MunchausenDQNAgent(AgentWithSimplePolicy):
+    """Munchausen DQN Agent based on PyTorch.
+
+    Notes
+    -----
+    Uses Munchausen trick for DQN for computing targets by default.
```
Comment (TimotheeMathieu) [inline rlberry/agents/torch/dqn/mdqn.py:50]: Can you add a word saying why one may want to use M-DQN instead of DQN ?

File: rlberry/agents/torch/dqn/dqn_utils.py [hunk 2]
```diff
@@ -8,6 +10,38 @@
 logger = rlberry.logger
 
 
+def stable_scaled_log_softmax(x, tau, axis=-1):
+    """Scaled log_softmax operation.
+    Args:
```
Comment (TimotheeMathieu) [inline rlberry/agents/torch/dqn/dqn_utils.py:15]: `Parameters `instead of `Args` ? the is the convention we have.

File: rlberry/agents/torch/dqn/dqn_utils.py [hunk 3]
```diff
@@ -8,6 +10,38 @@
...
+    y = x - max_x
+    tau_lse = max_x + tau * torch.log(
+        torch.sum(torch.exp(y / tau), dim=axis, keepdim=True)
+    )
+    return x - tau_lse
+
+
+def stable_softmax(x, tau, axis=-1):
+    """Stable softmax operation.
+    Args:
```
Comment (TimotheeMathieu) [inline rlberry/agents/torch/dqn/dqn_utils.py:32]: same as above

File: rlberry/agents/torch/dqn/dqn_utils.py (hunk 2 above)
Comment (TimotheeMathieu) [reply rlberry/agents/torch/dqn/dqn_utils.py:15]: and with the usual formatting with ----------- below the parameters and returns keywords. This is if one day we want to include this in the API and also to have somthing well formated when asking the docstring directly from python

**Review by TimotheeMathieu — COMMENTED** [review]
LGTM also, I added a few comments. And I agree that the example you put is sufficient for first PR and you can do another PR for comparison with other algo.

**Comment by TimotheeMathieu** [comment]
I added the label "ready for review" so that the test are done next time you commit. (next time you can do it with the buttons on the right of the screen :) )

**Review by TimotheeMathieu — APPROVED** [review]
LGTM, Thanks
